adagrams/game.py

```diff
@@ -1,11 +1,126 @@
+import random
+
 def draw_letters():
-    pass
+    # Import random - assign the randomized number to the index
+    # "bag" of letters will be a dictionary as shown below
+    # hand will be a list
+# | A : 9  | N : 6 |
+# | B : 2  | O : 8 |
+# | C : 2  | P : 2 |
+# | D : 4  | Q : 1 |
+# | E : 12 | R : 6 |
+# | F : 2  | S : 4 |
+# | G : 3  | T : 6 |
+# | H : 2  | U : 4 |
+# | I : 9  | V : 2 |
+# | J : 1  | W : 2 |
+# | K : 1  | X : 1 |
+# | L : 4  | Y : 2 |
+# | M : 2  | Z : 1 |
+    bag_of_letters_dict = {
+    'A': 9, 
+    'B': 2, 
+    'C': 2, 
+    'D': 4, 
+    'E': 12, 
+    'F': 2, 
+    'G': 3, 
+    'H': 2, 
+    'I': 9, 
+    'J': 1, 
+    'K': 1, 
+    'L': 4, 
+    'M': 2, 
+    'N': 6, 
+    'O': 8, 
+    'P': 2, 
+    'Q': 1, 
+    'R': 6, 
+    'S': 4, 
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+    bag_of_letters_list = []
+    for key, value in bag_of_letters_dict.items():
+        for letter in range(value):
+            bag_of_letters_list.append(key)
+
+    hand = []
+
+    while len(hand) < 10:
+        for letter in bag_of_letters_list:
+            hand.append(letter)
+            letter_of_index = bag_of_letters_list.index(letter)
+            bag_of_letters_list.pop(letter_of_index)
+            break
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    true_false = []
+    word = word.upper()
+    word_list = []
+
+    for letter in word:
+        word_list.append(letter)
+        if letter in letter_bank and word_list.count(letter) == letter_bank.count(letter):
+            true_false.append(True)
+        else:
+            true_false.append(False)
+    if False in true_false:
+        return False
+    else:
+        return True
 
 def score_word(word):
-    pass
+    scoreboard = {
+        1 : "A, E, I, O, U, L, N, R, S, T",
+        2 : "D, G",
+        3 : "B, C, M, P",
+        4 : "F, H, V, W, Y",
+        5 : "K",
+        8 : "J, X",
+        10 : "Q, Z"
+    }
+    word = word.upper()
+    #- If the length of the word is 7, 8, 9, or 10, 
+    # then the word gets an additional 8 points
+    score = 0
+    for letter in word:
+        for key, value in scoreboard.items():
+            if letter in value:
+                score += key
```

Comment:
This is nominally an O(n) solution, but only because the inner loop is not growing in relation to the input size.  For every letter in `word`, this solution will have to loop through all of the letters in `scoreboard` in order to find the right key-value pair.  This doesn't take advantage of dictionary lookup as an O(1) operation.  Using a `scoreboard` dictionary where each letter is a key and the value is the point value would result in a more efficient solution, because then there would be no need for an inner loop (scoreboard[letter] would return the value for the letter).  In the worst case scenario, if all of the letters of word are 'Z', the inner loop in this solution will have to do 26 comparison operations, but if scoreboard is set up as {'A':1, 'B':3 ... 'Z':10}, the inner loop is not needed and there's only one operation needed to find the value for the letter.